core/lib/Cardano/Address/Style/Icarus.hs

```diff
@@ -218,8 +219,8 @@ instance Internal.HardDerivation Icarus where
 
     deriveAddressPrivateKey (Icarus accXPrv) accountingStyle addrIx =
         let
-            changeCode =
-                toEnum @(Index 'Soft _) $ fromEnum accountingStyle
+            changeCode = toEnum @(Index 'Soft _) $
+                fromEnum (invariantAccountingStyle accountingStyle)
```

Comment:
I don't think that's an acceptable way to do this. This will throw an exception in a public interface (users can provide `accountingStyle` and are now able to trigger some partial code).

I'd suggest to either:

- Have two distinguish types: 

```hs 
data Role 
    = UTxOExternal
    | UTxOInternal
    | Stake
    | Multisig 

newtype IsChange = IsChange { isChange :: Bool } 
```

And have the exposed interface use `IsChange` which is then mapped internally to the right `Role`. This way, the API remains type-safe and we don't compromise the flexibility of `deriveXPrv` and `deriveXPub`

Reply:
Opted for separated Role for each Jormungandr, Icarus and Shelley, removed AccountingStyle from Derivation